File: packages-ts/weather/src/live/sources.ts
// Phase 11 — per-source registry + polite floors.
//
// Mirrors Python `tradewinds/live/_sources.py`. Source set + floor values are
// kept byte-faithful with Python so cross-SDK consumers see the same
// invariant. Any divergence triggers the dual-SDK parity gate.

/** Canonical ordered tuple of supported sources. Order matters — keep AWC first. */
export const SUPPORTED_SOURCES = ["awc", "iem"] as const;

/** Validated source enum derived from `SUPPORTED_SOURCES`. */
export type LiveSource = (typeof SUPPORTED_SOURCES)[number];

/**
 * Minimum allowed poll cadence per source, in seconds.
 *
 * - AWC: 30s — aviationweather.gov has no documented rate limit but 30s is
 *   the empirically-validated floor that won't trip anti-abuse heuristics.
 * - IEM: 60s — mesonet.agron.iastate.edu is a university server; IEM docs
 *   explicitly ask for reasonable headroom above 1 req/s.
 */
export const POLITE_FLOORS_S: Readonly<Record<LiveSource, number>> = {
  awc: 30,
  iem: 60,
};

/**
 * Canonical per-source `source` field tag emitted on every observation row.
 *
 * `"awc.live"` / `"iem.live"` are the live-channel identity tags — distinct
 * from the archive-channel `"awc"` / `"iem"` written by the historical
 * fetchers. Cross-SDK parity: these match Python `SOURCE_IDENTITY_TAGS`.
 */
export const SOURCE_IDENTITY_TAGS = {
  awc: "awc.live",
  iem: "iem.live",
} as const satisfies Readonly<Record<LiveSource, string>>;

export type LiveSourceTag = (typeof SOURCE_IDENTITY_TAGS)[LiveSource];

/**
 * Normalize and validate a `source` option.
 *
 * @param source - Caller-supplied source string. `undefined`/`null` defaults
 *   to the first entry in `SUPPORTED_SOURCES` (AWC). Case-insensitive.
 * @returns The normalized lowercase source name (one of `SUPPORTED_SOURCES`).
 * @throws `Error` when the source is not in `SUPPORTED_SOURCES`.
 */
export function validateSource(source: string | null | undefined): LiveSource {
  if (source === undefined || source === null) {
    return SUPPORTED_SOURCES[0];
  }
  const normalized = source.trim().toLowerCase();
  if (!isLiveSource(normalized)) {
    throw new Error(
      `unknown live source ${JSON.stringify(source)}; supported: ${JSON.stringify(
        SUPPORTED_SOURCES,
      )}`,
    );
  }
  return normalized;
}

/** Type guard: narrow a string to `LiveSource`. */
export function isLiveSource(s: string): s is LiveSource {
  return (SUPPORTED_SOURCES as ReadonlyArray<string>).includes(s);
}

/**
 * Apply the polite-floor invariant to a caller-supplied cadence.
 *
 * @param pollSeconds - Caller-supplied cadence. `undefined`/`null` → use the floor.
 * @param source - A *validated* source name (call `validateSource` first).
 * @returns The cadence to use, in seconds.
 * @throws `Error` when `pollSeconds` is below the polite floor.
 */
export function validatePollSeconds(
  pollSeconds: number | null | undefined,
  source: LiveSource,
): number {
  const floor = POLITE_FLOORS_S[source];
  if (pollSeconds === undefined || pollSeconds === null) {
    return floor;
  }
  if (pollSeconds < floor) {
    throw new Error(
      `pollSeconds=${pollSeconds} below polite floor ${floor}s for source=${JSON.stringify(
        source,
      )}`,
    );
  }
  return pollSeconds;
}

/** Map a validated source name to its canonical row-level identity tag. */
export function sourceTag(source: LiveSource): LiveSourceTag {
  return SOURCE_IDENTITY_TAGS[source];
}

File: packages/core/src/tradewinds/live/_sources.py
"""Phase 11 — per-source registry + polite floors.

Each upstream public API has its own politeness contract. We hard-code the
minimum-allowed `poll_seconds` per source so callers cannot accidentally
hammer the endpoint. Values below the floor raise ValueError at stream()
startup, not on the first tick — fail fast.

AWC: 30s floor. The aviationweather.gov live endpoint serves recent METARs
    with no documented rate limit; 30s is the floor we've validated empirically
    won't trip anti-abuse heuristics.
IEM: 60s floor. mesonet.agron.iastate.edu is a university server; the IEM
    documentation explicitly requests 1+ req/s with reasonable headroom.
    A 60s cadence comfortably stays below their threshold.
"""

from __future__ import annotations

from typing import Final

# Canonical ordered tuple of supported sources. ORDER MATTERS for the
# default-source resolution in stream()/latest() and for error-message
# stability — keep AWC first.
SUPPORTED_SOURCES: Final[tuple[str, ...]] = ("awc", "iem")

# Minimum allowed poll cadence per source. Streams that pass `poll_seconds`
# below the floor raise ValueError at startup. Streams that omit `poll_seconds`
# default to the floor for the active source.
POLITE_FLOORS_S: Final[dict[str, float]] = {
    "awc": 30.0,
    "iem": 60.0,
}

# Canonical per-source `source` field tag emitted on every observation row.
SOURCE_IDENTITY_TAGS: Final[dict[str, str]] = {
    "awc": "awc.live",
    "iem": "iem.live",
}


def validate_source(source: str | None) -> str:
    """Normalize and validate a source kwarg.

    Args:
        source: Caller-supplied source string. ``None`` → defaults to first
            entry in ``SUPPORTED_SOURCES`` (AWC). Case-insensitive.

    Returns:
        The normalized lowercase source name (one of ``SUPPORTED_SOURCES``).

    Raises:
        ValueError: When the source is not in ``SUPPORTED_SOURCES``.
    """
    if source is None:
        return SUPPORTED_SOURCES[0]
    normalized = source.strip().lower()
    if normalized not in SUPPORTED_SOURCES:
        raise ValueError(
            f"unknown live source {source!r}; supported: {SUPPORTED_SOURCES!r}"
        )
    return normalized


def validate_poll_seconds(poll_seconds: float | None, source: str) -> float:
    """Apply the polite-floor invariant to a caller-supplied cadence.

    Args:
        poll_seconds: Caller-supplied poll cadence. ``None`` → use the floor.
        source: A *validated* source name (call ``validate_source`` first).

    Returns:
        The cadence to use, in seconds.

    Raises:
        ValueError: When ``poll_seconds`` is below the polite floor.
    """
    floor = POLITE_FLOORS_S[source]
    if poll_seconds is None:
        return floor
    if poll_seconds < floor:
        raise ValueError(
            f"poll_seconds={poll_seconds} below polite floor "
            f"{floor}s for source={source!r}"
        )
    return float(poll_seconds)


def source_tag(source: str) -> str:
    """Map a validated source name to its canonical row-level identity tag."""
    return SOURCE_IDENTITY_TAGS[source]


__all__ = [
    "POLITE_FLOORS_S",
    "SOURCE_IDENTITY_TAGS",
    "SUPPORTED_SOURCES",
    "source_tag",
    "validate_poll_seconds",
    "validate_source",
]
